clash-prelude/src/Clash/Class/Resize.hs

```diff
@@ -117,3 +120,57 @@ checkedTruncateB ::
   f (a + b) -> f a
 checkedTruncateB v =
   checkIntegral (Proxy @(f a)) v `seq` truncateB v
+
+-- | Helper function of 'maybeFromIntegral', 'maybeResize' and 'maybeTruncateB'
+maybeIntegral ::
+  forall a b.
+  (Integral a, Integral b, Bounded b) =>
+  Proxy b ->
+  a -> Bool
+maybeIntegral Proxy v =
+     toInteger v < toInteger (maxBound @b)
+  && toInteger v > toInteger (minBound @b)
+
+-- | Like 'fromIntegral', but returns 'Nothing' if /a/ is out of bounds for /b/.
+-- Useful when you do not know /a/ can be out of bounds, and would like to have
+-- your assumptions checked.
+
+-- * __NB__: 'fromIntegral' is not well suited for Clash as it will go through
+-- 'Integer' which is arbitrarily bounded in HDL. Instead use
+-- 'Clash.Class.BitPack.bitCoerce' and the 'Resize' class.
+maybeFromIntegral ::
+  forall a b.
+  (Integral a, Integral b, Bounded b) =>
+  a -> Maybe b
+maybeFromIntegral v =
+  if maybeIntegral (Proxy @b) v
+    then Just (fromIntegral v)
+    else Nothing
+
+-- | Like 'resize', but returns 'Nothing' if /f a/ is out of bounds for /f b/.
+-- Useful when you do not know /f a/ can be out of bounds, and would like to
+-- have your assumptions checked.
```

Comment:
I think all these `maybe` variants can be described much shorter:

> Like `resize`, but returns `Nothing` if _f a_ is out of bounds for _f b_.

The rest seems unnecessary, and also it mentions assumptions, but we don't go in with assumptions. This function is total; the assumption with the `checked` variant was that it would fit, but we want to error when that assumption is wrong. This is different.

Also, Martijn and I discussed that we'd best rephrase all these functions a bit, but that can be done in a followup PR. Or I can tack a commit onto this PR...